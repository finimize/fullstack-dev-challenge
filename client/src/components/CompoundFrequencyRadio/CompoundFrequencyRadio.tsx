import React, { FC, useContext } from 'react'
import { Box, useRadio, UseRadioProps, useRadioGroup, FormLabel } from '@chakra-ui/react'
import { AppContext } from '../../store'
import { TOGGLE_COMPOUNDING_FREQUENCY } from '../../store/types'
import { CompoundingFrequency } from '../../store/store.interface'

interface RadioCardProps extends UseRadioProps {
    children: React.ReactNode
}

export const RadioCard: FC<RadioCardProps> = (props) => {
    const { getInputProps, getCheckboxProps } = useRadio(props)

    const input = getInputProps()
    const checkbox = getCheckboxProps()
    const { value } = props
    return (
        <Box display='grid' as='label' justifyContent='center'>
            <input {...input} data-testid={`cfr-${value as string}`} />
            <Box
                {...checkbox}
                cursor='pointer'
                borderWidth='1px'
                borderRadius='md'
                boxShadow='md'
                width='fit-content'
                bg='white'
                _checked={{
                    bg: 'blue500',
                    color: 'white',
                    borderColor: 'white',
                    borderWidth: '2px',
                }}
                _focus={{
                    boxShadow: 'outline',
                }}
                px={2}
                py={2}
                textAlign='center'
                fontSize='xs'
                fontWeight='700'
                margin='2'
                height='fit-content'
            >
                {props.children}
            </Box>
        </Box>
    )
}

export const CompoundFrequencyRadio: FC = () => {
    const { state, dispatch } = useContext(AppContext)
    const options = [
        CompoundingFrequency[CompoundingFrequency.annually],
        CompoundingFrequency[CompoundingFrequency.quarterly],
        CompoundingFrequency[CompoundingFrequency.monthly],
    ]
    const { getRootProps, getRadioProps } = useRadioGroup({
        name: 'framework',
        defaultValue: CompoundingFrequency[state.calculations.compoundingFrequency],
        onChange: (value: string) =>
            dispatch({ type: TOGGLE_COMPOUNDING_FREQUENCY, payload: value }),
    })

    const group = getRootProps()
    return (
        <Box padding='2' marginBottom='2'>
            <FormLabel fontSize='sm' marginBottom='2' color='text' fontWeight='500'>
                Interest Frequency
            </FormLabel>
            <Box
                {...group}
                display='grid'
                py='9px'
                height='auto'
                width='100%'
                backgroundColor='grey3'
                borderRadius='8px'
                justifyContent='space-evenly'
                alignItems='center'
                // height='72px'
                gridTemplateColumns={{
                    base: '1fr',
                    sm: 'repeat(3,1fr)',
                }}
            >
                {options.map((value) => {
                    const radio = getRadioProps({ value })
                    return (
                        <RadioCard key={value} {...radio}>
                            {value}
                        </RadioCard>
                    )
                })}
            </Box>
        </Box>
    )
}
