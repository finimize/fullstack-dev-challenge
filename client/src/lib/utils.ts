import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
