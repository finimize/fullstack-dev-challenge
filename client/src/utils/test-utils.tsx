import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
