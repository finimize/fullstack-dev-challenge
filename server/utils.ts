const compoundedInterest = (interestRate: number, timeInMonths: number) :number => {
    return Math.pow((interestRate + 1), timeInMonths);
};

const percentageToDecimal = (percentValue: number) :number => percentValue/100;

export const monthlyCompoundInterest = (monthlyDeposit: number, monthlyInterest: number, timeInMonths: number) :number =>  {
    const interestRate = percentageToDecimal(monthlyInterest);
    const interestOverMonths = compoundedInterest(interestRate, timeInMonths);
    const monthlyCompound = (monthlyDeposit *((interestOverMonths-1)/interestRate));
    return Math.round(monthlyCompound);
}

export const principleCompoundInterest = (initialDeposit: number,
                                          monthlyInterest: number,
                                          timeInMonths: number) : number => {
    const interestOverMonths = compoundedInterest(percentageToDecimal(monthlyInterest), timeInMonths);
    return Math.round(initialDeposit * interestOverMonths);
}

export const totalSavingsOverTime = (initialDeposit: number,
                                    monthlyDeposit: number,
                                    monthlyInterest: number,
                                    timeInMonths: number) :number => {
    const principle = principleCompoundInterest(initialDeposit, monthlyInterest, timeInMonths);
    const  monthly = monthlyCompoundInterest(monthlyDeposit, monthlyInterest, timeInMonths);
    return monthly+principle;
}

const toRange = (value: number) => {
    let array = [];
    for(let i =1;i<=value;i++){
        array.push(i);
    }
    return array;
}
export const toYearsFromNow = (years: number) :Array<string> => {
    let yearsArray = toRange(years);
    return yearsArray.map(number => {
        let d = new Date();
        let currentYear = d.getFullYear();
        let month = d.getMonth();
        let day = d.getDate();
        let c = new Date(currentYear + number, month, day);
        return c.getFullYear().toString();
    })
}
export const yearlySavingsOverTime = (initialDeposit: number,
                                        monthlyDeposit: number,
                                         monthlyInterest: number,
                                         timeInYears: number) : Array<number> => {
        const arr = toRange(timeInYears);
        const principleCompound = principleCompoundInterest(initialDeposit,monthlyInterest,timeInYears*12);
        return arr.map(number => monthlyCompoundInterest(monthlyDeposit,monthlyInterest,number*12) + principleCompound);
}
