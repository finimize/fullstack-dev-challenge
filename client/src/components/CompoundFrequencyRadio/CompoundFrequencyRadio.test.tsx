import React from 'react'
import { cleanup, fireEvent } from '@testing-library/react'
import { customRender } from '../../utils/test-utils'
import { CompoundFrequencyRadio } from './CompoundFrequencyRadio'

afterEach(cleanup)

describe('component/CompoundFrequencyRadio', () => {
    it('renders Calculator Mode component', () => {
        const { render } = customRender(<CompoundFrequencyRadio />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })

    it('allows compounding frequency mode to be changed', () => {
        const { render, dispatchMock } = customRender(<CompoundFrequencyRadio />, {})
        const { getByTestId } = render
        const annuallybutton = getByTestId('cfr-annually')
        const monthlybutton = getByTestId('cfr-monthly')
        const quarterlybutton = getByTestId('cfr-quarterly')
        // Check initial view
        expect(annuallybutton).toBeChecked()
        expect(monthlybutton).not.toBeChecked()
        expect(quarterlybutton).not.toBeChecked()
        // // Check monthly button is working
        fireEvent.click(monthlybutton)
        expect(dispatchMock).toHaveBeenCalledWith({
            payload: 'monthly',
            type: 'TOGGLE_COMPOUNDING_FREQUENCY',
        })
        expect(monthlybutton).toBeChecked()
        fireEvent.click(quarterlybutton)
        expect(dispatchMock).toHaveBeenNthCalledWith(2, {
            payload: 'quarterly',
            type: 'TOGGLE_COMPOUNDING_FREQUENCY',
        })
        expect(quarterlybutton).toBeChecked()
    })

    // it('allows calculator modes to be changed to simple mode', () => {
    //     const { render, dispatchMock } = customRender(<CalculatorMode />, {
    //         store: { ...initialState, calculatorMode: 'Advanced' },
    //     })
    //     const { getByText, getByTestId, container } = render
    //     expect(getByText('Simple')).toBeInTheDocument()
    //     const advancedbutton = getByTestId('Advanced')
    //     const simplebutton = getByTestId('Simple')
    //     // Check initial view
    //     expect(simplebutton).not.toHaveAttribute('checked')
    //     expect(advancedbutton).toBeChecked()
    //     // Check Advanced button is working
    //     fireEvent.click(simplebutton)
    //     expect(dispatchMock).toHaveBeenNthCalledWith(1, { payload: 'Simple', type: 'TOGGLE_MODE' })
    //     expect(dispatchMock).toHaveBeenNthCalledWith(2, {
    //         type: 'UPDATE_CALCULATIONS',
    //         field: 'monthlyContributions',
    //         payload: {
    //             value: '0',
    //         },
    //     })
    //     expect(dispatchMock).toHaveBeenNthCalledWith(3, {
    //         payload: 'annually',
    //         type: 'TOGGLE_COMPOUNDING_FREQUENCY',
    //     })
    //     expect(dispatchMock).toHaveBeenCalledTimes(3)
    //     expect(simplebutton).toBeChecked()
    //     expect(container).toMatchSnapshot()
    // })
    // it('allows to hit', () => {
    //     const { render, dispatchMock } = customRender(<CalculatorMode />, {
    //         store: { ...initialState, currentPage: 2 },
    //     })
    //     const { getByTestId } = render
    //     const prevbutton = getByTestId('previous-button')
    //     const nextbutton = getByTestId('next-button')
    //     userEvent.click(prevbutton)
    //     expect(dispatchMock).toHaveBeenCalledWith({ type: 'PREVIOUS_PAGE' })
    //     userEvent.click(nextbutton)
    //     expect(dispatchMock).toHaveBeenCalledWith({ type: 'NEXT_PAGE' })
    // })
})
