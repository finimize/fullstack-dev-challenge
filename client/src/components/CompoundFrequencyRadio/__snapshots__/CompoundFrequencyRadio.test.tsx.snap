// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/CompoundFrequencyRadio renders Calculator Mode component 1`] = `
<div>
   
  <div
    class="css-uarar0"
  >
    <label
      class="chakra-form__label css-wuh3rg"
    >
      Interest Frequency
    </label>
    <div
      class="css-171bmva"
      role="radiogroup"
    >
      <label
        class="css-l75mtb"
      >
        <input
          checked=""
          data-testid="cfr-annually"
          name="framework"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="radio"
          value="annually"
        />
        <div
          aria-hidden="true"
          class="css-15px1hi"
          data-checked=""
        >
          annually
        </div>
      </label>
      <label
        class="css-l75mtb"
      >
        <input
          data-testid="cfr-quarterly"
          name="framework"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="radio"
          value="quarterly"
        />
        <div
          aria-hidden="true"
          class="css-15px1hi"
        >
          quarterly
        </div>
      </label>
      <label
        class="css-l75mtb"
      >
        <input
          data-testid="cfr-monthly"
          name="framework"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="radio"
          value="monthly"
        />
        <div
          aria-hidden="true"
          class="css-15px1hi"
        >
          monthly
        </div>
      </label>
    </div>
  </div>
   
   
</div>
`;
